### ISSUE TITLE
Classpath Scanning in Spring Boot Jar Not Working

### ISSUE THREAD

**Issue by creactiviti** [lukehutch/fast-classpath-scanner#46]
I have a custom annotation:

```
@Target(value=ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
public @interface Pipeline  {
  String id ();
  String name ();
  boolean visible () default true;
  String role () default "USER";
}
```

Which I use on interfaces within a Spring Boot App with nested jars: 

```
@Pipeline(id="my-pipeline", name="My Pipeline")
public interface Echo {
  ...
}
```

My project structure:

```
parent-boot-project
   |
   --- plugin1.jar 
   --- plugin2.jar <--- interfaces are here
   --- ...
```

Next, I use lukehutch's fast-classpath-scanner to scan for them:

```
new FastClasspathScanner(BASE_PACKAGE)
    .matchClassesWithAnnotation(PIPELINE_ANNOTATION, aProcessor)
    .verbose()
    .scan();
```

Works great on my IDE (eclipse) but not when I build the jar. Tried messing around with the classpath but to no avail. Anyone ran into something like that?

**Comment by lukehutch** [comment]
Hi Arik, thanks for reporting this. I assume you're using the latest
release in Maven? If so, can you try with the head version in git, please?
This may have been fixed last week.

**Comment by lukehutch** [comment]
@creactiviti Any update on this? Did you test with the head version yet?

**Comment by lukehutch** [comment]
Fixed it 1.9.19.

**Comment by Dretch** [comment]
This still seems to be broken for me in `1.1.19` and `2.0.1`. Looking at the verbose output from the scan, it seems like scanner fails to find all the dependencies of the application, although it finds some of them:
```
FastClasspathScanner: Classpath elements from ClassLoader sun.misc.Launcher$AppClassLoader were extracted by ClassLoaderHandler io.github.lukehutch.fastclasspathscanner.classpath.classloaderhandler.URLClassLoaderHandler
FastClasspathScanner: Skipping JRE jar: /Library/Java/JavaVirtualMachines/jdk1.8.0_60.jdk/Contents/Home/jre/lib/ext/cldrdata.jar
FastClasspathScanner: Skipping JRE jar: /Library/Java/JavaVirtualMachines/jdk1.8.0_60.jdk/Contents/Home/jre/lib/ext/dnsns.jar
FastClasspathScanner: Skipping JRE jar: /Library/Java/JavaVirtualMachines/jdk1.8.0_60.jdk/Contents/Home/jre/lib/ext/jaccess.jar
FastClasspathScanner: Skipping JRE jar: /Library/Java/JavaVirtualMachines/jdk1.8.0_60.jdk/Contents/Home/jre/lib/ext/jfxrt.jar
FastClasspathScanner: Skipping JRE jar: /Library/Java/JavaVirtualMachines/jdk1.8.0_60.jdk/Contents/Home/jre/lib/ext/localedata.jar
FastClasspathScanner: Skipping JRE jar: /Library/Java/JavaVirtualMachines/jdk1.8.0_60.jdk/Contents/Home/jre/lib/ext/nashorn.jar
FastClasspathScanner: Skipping JRE jar: /Library/Java/JavaVirtualMachines/jdk1.8.0_60.jdk/Contents/Home/jre/lib/ext/sunec.jar
FastClasspathScanner: Skipping JRE jar: /Library/Java/JavaVirtualMachines/jdk1.8.0_60.jdk/Contents/Home/jre/lib/ext/sunjce_provider.jar
FastClasspathScanner: Skipping JRE jar: /Library/Java/JavaVirtualMachines/jdk1.8.0_60.jdk/Contents/Home/jre/lib/ext/sunpkcs11.jar
FastClasspathScanner: Skipping JRE jar: /Library/Java/JavaVirtualMachines/jdk1.8.0_60.jdk/Contents/Home/jre/lib/ext/zipfs.jar
FastClasspathScanner: Found classpath element: /opt/subversion/lib/libsvnjavahl-1.0.dylib
FastClasspathScanner: Found classpath element: /System/Library/Java/Extensions/AppleScriptEngine.jar
FastClasspathScanner: Found classpath element: /System/Library/Java/Extensions/dns_sd.jar
FastClasspathScanner: Found classpath element: /System/Library/Java/Extensions/j3daudio.jar
FastClasspathScanner: Found classpath element: /System/Library/Java/Extensions/j3dcore.jar
FastClasspathScanner: Found classpath element: /System/Library/Java/Extensions/j3dutils.jar
FastClasspathScanner: Found classpath element: /System/Library/Java/Extensions/jai_codec.jar
FastClasspathScanner: Found classpath element: /System/Library/Java/Extensions/jai_core.jar
FastClasspathScanner: Found classpath element: /System/Library/Java/Extensions/libAppleScriptEngine.jnilib
FastClasspathScanner: Found classpath element: /System/Library/Java/Extensions/libJ3D.jnilib
FastClasspathScanner: Found classpath element: /System/Library/Java/Extensions/libJ3DAudio.jnilib
FastClasspathScanner: Found classpath element: /System/Library/Java/Extensions/libJ3DUtils.jnilib
FastClasspathScanner: Found classpath element: /System/Library/Java/Extensions/libmlib_jai.jnilib
FastClasspathScanner: Found classpath element: /System/Library/Java/Extensions/mlibwrapper_jai.jar
FastClasspathScanner: Found classpath element: /System/Library/Java/Extensions/MRJToolkit.jar
FastClasspathScanner: Found classpath element: /System/Library/Java/Extensions/vecmath.jar
FastClasspathScanner: Found classpath element: /usr/lib/java/libjdns_sd.jnilib
FastClasspathScanner: Classpath elements from ClassLoader sun.misc.Launcher$ExtClassLoader were extracted by ClassLoaderHandler io.github.lukehutch.fastclasspathscanner.classpath.classloaderhandler.URLClassLoaderHandler
FastClasspathScanner: Classpath elements from ClassLoader org.springframework.boot.loader.LaunchedURLClassLoader were extracted by ClassLoaderHandler io.github.lukehutch.fastclasspathscanner.classpath.classloaderhandler.URLClassLoaderHandler
FastClasspathScanner: Classpath elements from ClassLoader org.springframework.boot.context.embedded.tomcat.TomcatEmbeddedWebappClassLoader were extracted by ClassLoaderHandler io.github.lukehutch.fastclasspathscanner.classpath.classloaderhandler.URLClassLoaderHandler
FastClasspathScanner: Classpath elements: [/Users/garethsmith/Documents/IDEA_projects/garethproject/webapp/target/webapp-1.0-SNAPSHOT.jar, /opt/subversion/lib/libsvnjavahl-1.0.dylib, /System/Library/Java/Extensions/AppleScriptEngine.jar, /System/Library/Java/Extensions/dns_sd.jar, /System/Library/Java/Extensions/j3daudio.jar, /System/Library/Java/Extensions/j3dcore.jar, /System/Library/Java/Extensions/j3dutils.jar, /System/Library/Java/Extensions/jai_codec.jar, /System/Library/Java/Extensions/jai_core.jar, /System/Library/Java/Extensions/libAppleScriptEngine.jnilib, /System/Library/Java/Extensions/libJ3D.jnilib, /System/Library/Java/Extensions/libJ3DAudio.jnilib, /System/Library/Java/Extensions/libJ3DUtils.jnilib, /System/Library/Java/Extensions/libmlib_jai.jnilib, /System/Library/Java/Extensions/mlibwrapper_jai.jar, /System/Library/Java/Extensions/MRJToolkit.jar, /System/Library/Java/Extensions/vecmath.jar, /usr/lib/java/libjdns_sd.jnilib]
FastClasspathScanner: *** Starting scan ***
FastClasspathScanner: => Scanning classpath element: /Users/garethsmith/Documents/IDEA_projects/garethproject/webapp/target/webapp-1.0-SNAPSHOT.jar
FastClasspathScanner: Scanning jarfile: /Users/garethsmith/Documents/IDEA_projects/garethproject/webapp/target/webapp-1.0-SNAPSHOT.jar
FastClasspathScanner: Found class com.garethproject.runtime.config.ContextConfiguration
FastClasspathScanner: Class com.garethproject.runtime.config.ContextConfiguration has annotation org.springframework.context.annotation.Configuration
FastClasspathScanner: Class com.garethproject.runtime.config.ContextConfiguration has annotation org.springframework.context.annotation.ImportResource
FastClasspathScanner: Found class com.garethproject.runtime.config.EmbeddedServletContainerConfiguration with superclass org.springframework.web.servlet.config.annotation.WebMvcConfigurerAdapter
FastClasspathScanner: Class com.garethproject.runtime.config.EmbeddedServletContainerConfiguration implements interface org.springframework.boot.context.embedded.EmbeddedServletContainerCustomizer
FastClasspathScanner: Class com.garethproject.runtime.config.EmbeddedServletContainerConfiguration implements interface org.springframework.boot.context.embedded.ServletContextInitializer
FastClasspathScanner: Class com.garethproject.runtime.config.EmbeddedServletContainerConfiguration has a field with type or type parameter org.springframework.context.ApplicationContext
FastClasspathScanner: Class com.garethproject.runtime.config.EmbeddedServletContainerConfiguration has a field with type or type parameter org.springframework.core.env.Environment
FastClasspathScanner: Class com.garethproject.runtime.config.EmbeddedServletContainerConfiguration has annotation org.springframework.context.annotation.Configuration
FastClasspathScanner: Found class com.garethproject.runtime.config.HttpInvokerServiceConfiguration
FastClasspathScanner: Class com.garethproject.runtime.config.HttpInvokerServiceConfiguration has a field with type or type parameter com.garethproject.event.service.remote.EventImportRemoteService
FastClasspathScanner: Class com.garethproject.runtime.config.HttpInvokerServiceConfiguration has annotation org.springframework.context.annotation.Configuration
FastClasspathScanner: Found class com.garethproject.runtime.controller.ErrorController
FastClasspathScanner: Class com.garethproject.runtime.controller.ErrorController has a field with type or type parameter org.springframework.core.env.Environment
FastClasspathScanner: Class com.garethproject.runtime.controller.ErrorController has annotation org.springframework.stereotype.Controller
FastClasspathScanner: Found class com.garethproject.runtime.health.HazelcastHealthIndicator
FastClasspathScanner: Class com.garethproject.runtime.health.HazelcastHealthIndicator implements interface org.springframework.boot.actuate.health.HealthIndicator
FastClasspathScanner: Class com.garethproject.runtime.health.HazelcastHealthIndicator has a field with type or type parameter com.hazelcast.core.HazelcastInstance
FastClasspathScanner: Class com.garethproject.runtime.health.HazelcastHealthIndicator has a field with type or type parameter com.hazelcast.core.LifecycleEvent$LifecycleState
FastClasspathScanner: Class com.garethproject.runtime.health.HazelcastHealthIndicator has annotation org.springframework.stereotype.Component
FastClasspathScanner: Found class com.garethproject.WebDeployable
FastClasspathScanner: Class com.garethproject.WebDeployable has annotation org.springframework.boot.autoconfigure.SpringBootApplication
FastClasspathScanner: Scanned jarfile /Users/garethsmith/Documents/IDEA_projects/garethproject/webapp/target/webapp-1.0-SNAPSHOT.jar in 28 msec
FastClasspathScanner: => Scanning classpath element: /opt/subversion/lib/libsvnjavahl-1.0.dylib
FastClasspathScanner: => Scanning classpath element: /System/Library/Java/Extensions/AppleScriptEngine.jar
FastClasspathScanner: Scanning jarfile: /System/Library/Java/Extensions/AppleScriptEngine.jar
FastClasspathScanner: Scanned jarfile /System/Library/Java/Extensions/AppleScriptEngine.jar in 0 msec
FastClasspathScanner: => Scanning classpath element: /System/Library/Java/Extensions/dns_sd.jar
FastClasspathScanner: Scanning jarfile: /System/Library/Java/Extensions/dns_sd.jar
FastClasspathScanner: Scanned jarfile /System/Library/Java/Extensions/dns_sd.jar in 0 msec
FastClasspathScanner: => Scanning classpath element: /System/Library/Java/Extensions/j3daudio.jar
FastClasspathScanner: Scanning jarfile: /System/Library/Java/Extensions/j3daudio.jar
FastClasspathScanner: Scanned jarfile /System/Library/Java/Extensions/j3daudio.jar in 0 msec
FastClasspathScanner: => Scanning classpath element: /System/Library/Java/Extensions/j3dcore.jar
FastClasspathScanner: Scanning jarfile: /System/Library/Java/Extensions/j3dcore.jar
FastClasspathScanner: Scanned jarfile /System/Library/Java/Extensions/j3dcore.jar in 1 msec
FastClasspathScanner: => Scanning classpath element: /System/Library/Java/Extensions/j3dutils.jar
FastClasspathScanner: Scanning jarfile: /System/Library/Java/Extensions/j3dutils.jar
FastClasspathScanner: Scanned jarfile /System/Library/Java/Extensions/j3dutils.jar in 1 msec
FastClasspathScanner: => Scanning classpath element: /System/Library/Java/Extensions/jai_codec.jar
FastClasspathScanner: Scanning jarfile: /System/Library/Java/Extensions/jai_codec.jar
FastClasspathScanner: Scanned jarfile /System/Library/Java/Extensions/jai_codec.jar in 0 msec
FastClasspathScanner: => Scanning classpath element: /System/Library/Java/Extensions/jai_core.jar
FastClasspathScanner: Scanning jarfile: /System/Library/Java/Extensions/jai_core.jar
FastClasspathScanner: Scanned jarfile /System/Library/Java/Extensions/jai_core.jar in 2 msec
FastClasspathScanner: => Scanning classpath element: /System/Library/Java/Extensions/libAppleScriptEngine.jnilib
FastClasspathScanner: => Scanning classpath element: /System/Library/Java/Extensions/libJ3D.jnilib
FastClasspathScanner: => Scanning classpath element: /System/Library/Java/Extensions/libJ3DAudio.jnilib
FastClasspathScanner: => Scanning classpath element: /System/Library/Java/Extensions/libJ3DUtils.jnilib
FastClasspathScanner: => Scanning classpath element: /System/Library/Java/Extensions/libmlib_jai.jnilib
FastClasspathScanner: => Scanning classpath element: /System/Library/Java/Extensions/mlibwrapper_jai.jar
FastClasspathScanner: Scanning jarfile: /System/Library/Java/Extensions/mlibwrapper_jai.jar
FastClasspathScanner: Scanned jarfile /System/Library/Java/Extensions/mlibwrapper_jai.jar in 0 msec
FastClasspathScanner: => Scanning classpath element: /System/Library/Java/Extensions/MRJToolkit.jar
FastClasspathScanner: Scanning jarfile: /System/Library/Java/Extensions/MRJToolkit.jar
FastClasspathScanner: Scanned jarfile /System/Library/Java/Extensions/MRJToolkit.jar in 0 msec
FastClasspathScanner: => Scanning classpath element: /System/Library/Java/Extensions/vecmath.jar
FastClasspathScanner: Scanning jarfile: /System/Library/Java/Extensions/vecmath.jar
FastClasspathScanner: Scanned jarfile /System/Library/Java/Extensions/vecmath.jar in 0 msec
FastClasspathScanner: => Scanning classpath element: /usr/lib/java/libjdns_sd.jnilib
FastClasspathScanner: *** Time taken by .scan(): 38 ms ***
```

**Comment by lukehutch** [comment]
@Dretch thanks for the update. Are you working on the same project as @creactiviti , or is this a separate case?

Can you please list some of the jars that are not found? Is there something different about the jars that are found and the jars that are not, or about how they are added to the classpath?

Even better, if you could please create a minimal testcase and upload it here (or provide duplication instructions for a minimal testcase), I'll take a look. Thanks!

**Comment by Dretch** [comment]
Hi @lukehutch. My project is not connected with that of @creactiviti 

I have created a testcase, uploaded here: https://drive.google.com/file/d/0B2aUdXQz2bVMSnNKWG9IOTdwNUU/view?usp=sharing

Look at the `DemoApplication` class in `demo-subproject-deployable`. When this class is run as a spring-boot application from inside of IntelliJ then the `Fruit`/`Vegetable` subclasses are found correctly, but when the application is run from the jar file (`java -jar demo-subproject-deployable-0.0.1-SNAPSHOT.jar`) then the subclasses are not found.

Thanks for looking into this.

**Comment by lukehutch** [comment]
@Dretch thanks, I see what's happening here. Spring includes directories and jarfiles within jarfiles in the classpath, using the '!' URL separator, e.g.

```
demo-subproject-deployable-0.0.1-SNAPSHOT.jar!/BOOT-INF/classes
demo-subproject-deployable-0.0.1-SNAPSHOT.jar!/BOOT-INF/lib/spring-boot-starter-logging-1.4.0.RELEASE.jar
demo-subproject-deployable-0.0.1-SNAPSHOT.jar!/BOOT-INF/lib/slf4j-api-1.7.21.jar
...
```

This is not currently supported in FastClasspathScanner, and it's a bit complicated to support due to limitations in the `java.util.zip` API. I have a solution in mind, will try to fix this soon.

**Comment by Dretch** [comment]
Ah I see, cheers @lukehutch  👍

**Comment by lukehutch** [comment]
Hi @Dretch and @creactiviti, this bug should be fixed in 2.0.3. I would appreciate it if you could please each test this with your project, and let me know if it's fixed for you.

**Comment by lukehutch** [comment]
PS @Dretch, in your example code, note that you can add both the MatchProcessors to a single FastClasspathScanner instance, and perform the scan (for both fruit and vegetables) in a single pass.

**Comment by Dretch** [comment]
The fix worked for me. Thanks a bunch @lukehutch (and cheers for the usage heads-up).

**Comment by lukehutch** [comment]
@Dretch you're welcome, thanks for testing, and for taking the time to create a minimal testcase -- I couldn't have figured this out without it, and the scanning is a lot more powerful now that FCS supports scanning jars inside jars. (I had put that off for a long time, thinking nobody would ever run into it ;-) )

@creactiviti if this also fixes the issue for you, please go ahead and close the bug. Thank you for the report!

**Comment by lukehutch** [comment]
Closing since @Dretch confirmed this is fixed for him. @creactiviti, if you have a chance to test, please report your results here. Thank you!

**Comment by rmalchow** [comment]
i THINK this might be broken again - i just upgraded from 2.0.10 to 2.0.18,
and it stopped working for me.

i have a spring boot app that scans for velocity templates on the classpath.


```
	2.0.10 - OK
	2.0.11 - OK
	2.0.12 - OK
	2.0.13 and above:

		FastClasspathScanner	---- java.io.FileNotFoundException: /private/var/folders/hg/z_97z7494zb43q9mbbg459p40000gn/T/FastClasspathScanner-4706527374110542066---spring-webmvc-4.3.4.RELEASE.jar (No such file or directory)
		FastClasspathScanner	---- 	at java.util.zip.ZipFile.open(Native Method)
		FastClasspathScanner	---- 	at java.util.zip.ZipFile.<init>(ZipFile.java:219)
		FastClasspathScanner	---- 	at java.util.zip.ZipFile.<init>(ZipFile.java:149)
		FastClasspathScanner	---- 	at java.util.zip.ZipFile.<init>(ZipFile.java:120)
		FastClasspathScanner	---- 	at io.github.lukehutch.fastclasspathscanner.utils.NestedJarHandler$1$1.newInstance(NestedJarHandler.java:76)
		FastClasspathScanner	---- 	at io.github.lukehutch.fastclasspathscanner.utils.NestedJarHandler$1$1.newInstance(NestedJarHandler.java:73)
		FastClasspathScanner	---- 	at io.github.lukehutch.fastclasspathscanner.utils.Recycler.acquire(Recycler.java:50)
		FastClasspathScanner	---- 	at io.github.lukehutch.fastclasspathscanner.scanner.ClasspathElementZip.openInputStreamAndProcessFileMatch(ClasspathElementZip.java:273)
		FastClasspathScanner	---- 	at io.github.lukehutch.fastclasspathscanner.scanner.ClasspathElement.callFileMatchProcessors(ClasspathElement.java:283)
		FastClasspathScanner	---- 	at io.github.lukehutch.fastclasspathscanner.scanner.ScanSpec.callMatchProcessors(ScanSpec.java:477)
		FastClasspathScanner	---- 	at io.github.lukehutch.fastclasspathscanner.FastClasspathScanner.scan(FastClasspathScanner.java:1149)
		FastClasspathScanner	---- 	at io.github.lukehutch.fastclasspathscanner.FastClasspathScanner.scan(FastClasspathScanner.java:1206)
		FastClasspathScanner	---- 	at io.github.lukehutch.fastclasspathscanner.FastClasspathScanner.scan(FastClasspathScanner.java:1229) 
	
```
	
	
my code is:


```
	FastClasspathScanner fcs = new FastClasspathScanner();
	fcs.verbose(true); // just for debugging this
	
	fcs.matchFilenamePattern("templates/.*/.*\\.vm", new FileMatchContentsProcessor() {
		// ... process templates 
	}

	fcs.scan();

```

**Comment by lukehutch** [comment]
@rmalchow thanks for the report. Can you please include the whole verbose log output? (A pastebin link to the log would be great.)

**Comment by rmalchow** [comment]
hi luke,

the full output is ... er ... pretty long. 90+ MB uncompressed. tgz is ~ 1.5MB:

[out.txt.zip](https://github.com/lukehutch/fast-classpath-scanner/files/914612/out.txt.zip)

hope this helps

.rm

**Comment by lukehutch** [comment]
@rmalchow Yes, the reason for the logs being so big is that you are matching pattern `.*`, which matches a lot of files within your jarfiles :-)  (and there's the fact that an exception is thrown on every file, too!)

I'm checking in a change now that should fix this. Thank for catching this. Can you please test the git master version? If it works for you, I will push out a release. If it's too much trouble to test from git, I'll push out a release anyway, and you can grab it from Maven Central. Please let me know.

**Comment by rmalchow** [comment]
@lukehutch oops - yeah, i added that when i was trying to figure out what was going on. forgot to remove it. anyway. it seems to be doing what it's supposed to do. not sure about any side effects of course, but ... looking good. thank you tons for the quick response! awesome.

**Comment by lukehutch** [comment]
@rmalchow You're welcome. Should be fixed for good now, I had missed the case that you were triggering. Released in 2.0.19. Thanks again.

**Comment by CodingOX** [comment]
thanks..  use fast-classpath-scanner in springboot jar,   my problem got solved!!
